### ISSUE TITLE
obfs4 bridge doesn't start with custom option ExitNodes

### ISSUE THREAD

**Issue by Ahmed0425** [DcodingTheWeb/ProxAllium#3]
I've tried to use bridges with "ExitNodes" custom option in config.torrc and the result is that the process "obfs4proxy.exe" didn't start at all and connecting ended with dead connections during handshake (Oops, censorship firewall hit)

**Comment by TheDcoder** [comment]
I just tried this and I am getting these messages when I start Tor:
```
Failed to open GEOIP file Tor\geoip.  We've been configured to use (or avoid) nodes in certain countries, and we need GEOIP information to figure out which ones they are.
Failed to open GEOIP file Tor\geoip6.  We've been configured to use (or avoid) nodes in certain countries, and we need GEOIP information to figure out which ones they are.
```

Can you check if you are getting this message too? It should be the case most likely.

The reason for the error is that the `geoip` files which are used to map IP addresses to countries are missing. I intentionally decided to stop packing them into the default download package because 1) They were too big 2) Almost no one would use them without any GUI interface.

This can simply be fixed by downloading the Tor Expert Bundle ([`Download Tor`](https://www.torproject.org/download/download.html.en) -> `Windows` -> `Download Expert Bundle`) and extracting the files `geoip` and `geoip6` files to the `Tor` folder.

**Comment by Ahmed0425** [comment]
I figured out that the problem is that I should delete "tor data" folder and add geoip and geoip6 files
this solved the problem, although I got a couple of errors about general socks server failure (but it connected afterwards)
you should make a commit solving this "tor data" folder to avoid being deleting it manually every time I enter a different fingerprint
Is that "general server socks failure" error related to tor censorship?

**Comment by TheDcoder** [comment]
Interesting to see that `Tor Data` is causing trouble, that folder is used by Tor to store data about the network and it should normally don't cause any problems, I will check with the Tor folks to look into this problem... and also maybe add a "Clear Tor Data (Cache)" option while I am at it.

I am also getting the "general SOCKS server failure" warning, it is from Tor and I am currently investigating the error.

**Comment by Ahmed0425** [comment]
I've just tried 9150 port (found it in tor browser settings) instead of 9050 and it seems to be fixing a part of the "general socks server failure" issue as I encountered a couple of it only once during connection and didn't encounter it after making the circuit

**Comment by TheDcoder** [comment]
I do not recommend using the port 9150 as it is the default port used by the Tor Browser Bundle, 9050 is Tor's default port hence it is used by ProxAllium too.

I just tried a few things with bridges. I tested with both port 9150 and 9050 (they shouldn't make a difference) and the bridges didn't work with both of the ports, after grabbing a new pair of bridges I got no errors or warnings and I was connected to the Tor network normally via obfs4.

So I think this was a case of some broken/rouge relay misbehaving and causing the error, try changing the set of bridges that you are using and repeat the process until you don't get the error. It worked for me with the 3rd set of new bridges that I tried.

The port doesn't make a difference since it is purely local and not used for communication outside the computer.

**Comment by TheDcoder** [comment]
I am closing this thread because the main topic has been taken care of. I am successfully able to use custom exit nodes with obfs4 bridges.

Feel free to make a new issue or talk with me on IRC (details on the Chat page in the website) 😄

**Comment by F50Ci37Y** [comment]
> I am closing this thread because the main topic has been taken care of. I am successfully able to use custom exit nodes with obfs4 bridges.
> 
> Feel free to make a new issue or talk with me on IRC (details on the Chat page in the website) 😄

@TheDcoder 
How set ExitNodes?

**Comment by TheDcoder** [comment]
@fakhamatia You can set `ExitNodes` by editing the `config.torrc` file, open it in a text editor and you should see that there is a section at the bottom where you can specify custom settings :slightly_smiling_face:

For further information consult [Tor manual](https://2019.www.torproject.org/docs/tor-manual.html.en), just search for `ExitNodes` and you will find the relavent documentation!

**Comment by F50Ci37Y** [comment]
I try editing the config.torrc file but Tor did not connect stuck in 50% 
![Annotation 2020-01-23 191845](https://user-images.githubusercontent.com/20030910/73000371-027a6600-3e16-11ea-85df-1613091a850f.jpg)

**Comment by TheDcoder** [comment]
Are you using the latest version of ProxAllium along with the latest version of Tor Browser? Pay attention to the note on the download page:

> Note for previous users: If you are a previous user of ProxAllium and are updating now, I encourage you to install Tor Browser and use Tor from it as Tor is no longer bundled with ProxAllium (see [changelog]). The process is simple and ProxAllium will guide you through most of it, to get started with the Tor update, delete the "Tor" folder before running the updated version.

**Comment by F50Ci37Y** [comment]
Yes, last version ProxAllium and Tor.
I install the Tor Browser and copy the Tor folder to the ProxAllium folder.
It connects successfully to Tor without ExitNodes.

**Comment by TheDcoder** [comment]
> I install the Tor Browser and copy the Tor folder to the ProxAllium folder.

Please don't copy the Tor folder to ProxAllium, instead I recommend letting ProxAllium figure the folder location by using the semi-automated setup guide. You can trigger this guide by deleting your Tor folder in ProxAllium folder.

If this still doesn't solve the issue, I am afraid there is nothing much ProxAllium can do about it, it would be a general Tor issue.
